### ISSUE TITLE
Dual-source blending doesn't work correctly on Panfrost

### ISSUE THREAD

**Issue by icecream95** [hrydgard/ppsspp#15413]
### Game or games this happens in

Most of them…

### What area of the game / PPSSPP

Use the GLES renderer on a GPU supporting dual-source blending and with a renderer string including "Mali".

This occurs when using the Panfrost OpenGL driver on any Mali GPU, but setting the environment variable `force_gl_renderer=Mali` on any other system using Mesa for OpenGL drivers should also make the issue visible.

Because PPSSPP disables the `GLES3` flag for renderer strings which include "Mali", this snippet of code in `GLQueueRunner.cpp` does not execute the body of the `if` statement, and so `fragColor1` is written to render target 1 rather than the dual-source output for render target 0, which according to the OpenGL specification causes undefined behaviour.

```
                        if (gl_extensions.GLES3 && step.create_program.support_dual_source) {
                                glBindFragDataLocationIndexedEXT(program->program, 0, 0, "fragColor0");
                                glBindFragDataLocationIndexedEXT(program->program, 0, 1, "fragColor1");
                        }
```

Most games (user reports include "Ratchet Clank" and Tekken 6) show glitches, because the blend shader ends up reading a random value for the dual-source output from the fragment shader:

![snap0000266993](https://user-images.githubusercontent.com/29191311/154890703-6b512e50-4f82-4e25-aa45-b274f121be21.png)

### What should happen

- The `GLES3` flag should only get disabled for the proprietary ARM driver, by checking `GL_VENDOR` against "ARM" rather than using the renderer string. Even if this does expose some bugs in Panfrost, they can be fixed, and that is still better than the current behaviour.

- The code in `GLQueueRunner.cpp` should call `glBindFragDataLocationIndexedEXT` whenever dual-source blending is used, even if `GLES3` is not set.

### Platform

Linux / BSD

### Mobile phone model or graphics card

Mali G72

### PPSSPP version affected

v1.12.3

### Graphics backend (3D API)

OpenGL / GLES

### Checklist

- [X] Test in the [latest git build](https://buildbot.orphis.net/ppsspp/) in case it's already fixed.
- [X] [Search for other reports](https://github.com/hrydgard/ppsspp/search?q=my+issue&type=issues) of the same issue.
- [X] Try resetting settings or older versions and include if the issue is related.
- [X] Try without any cheats and without loading any save states.
- [X] Include logs or screenshots of issue.

**Comment by unknownbrackets** [comment]
Hm, see #4078.  This seems related to that issue on the forum, here:
https://forums.ppsspp.org/showthread.php?tid=24794

There, someone suggested this might also be related to highp/mediump.  I suppose that bug has been fixed, though:
https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/13714

I've created #15414, but it won't fix your issue alone, because we still force disable dual source blending for anything detected as GLES2 (because all the drivers are so buggy.)  And it seems like we're not detecting your driver as GLES3.  That said, you can try setting `VendorBugChecksEnabled = False` in ppsspp.ini to confirm if that's the only problem, once #15414 is merged.

We're using the following checks to detect GLES3:
 * `glGetIntegerv(GL_MAJOR_VERSION, &value)` should return 3, as per the spec.
 * `glGetError()` should return `GL_NO_ERROR` after getting `GL_MAJOR_VERSION` and `GL_MINOR_VERSION`.
 * The first digits in `glGetString(GL_VERSION)` should match the major and minor versions.
 * `glGetString(GL_VERSION)` should contain the characters `"OpenGL ES"` at the start.

If any of the above fail, we still look for 3.0 or 3.1 in the `glGetString(GL_VERSION)` string, but we block it for Mali since apparently that had 3.0 or 3.1 in some non-GLES3 driver version.

Can you confirm what this driver is returning for the following?
 * `GL_MAJOR_VERSION`
 * `GL_MINOR_VERSION`
 * `GL_VERSION`

-[Unknown]

**Comment by icecream95** [comment]
https://gitlab.com/panfork/mesa/-/commit/58eaf6c1f20d427b9590d24cc666753fe6a6029b

**Comment by icecream95** [comment]
> And it seems like we're not detecting your driver as GLES3

Hey, I didn't write *all* of it, only a few hundred patches here and there.

Saying that it was caused by the `strstr` against "Mali" was just an assumption from reading the source, I don't have any evidence that it actually was because of this.

I actually don't have PPSSPP built myself, and I was just looking at an [Apitrace](https://apitrace.github.io) from a user. In the trace file, GLX contexts are created and deleted multiple times until finally we (I assume) create a GLES context with EGL and use that. (The trace does not record any EGL functions at all.)

```
7 glXChooseVisual(dpy = 0xaaaae035a560, screen = 0, attribList = {GLX_RGBA, GLX_RED_SIZE, 3, GLX_GREEN_SIZE, 3, GLX_BLUE_SIZE, 2, GLX_DOUBLEBUFFER, GLX_DEPTH_SIZE, 16, 0}) = &{visual = 0xaaaae03687f8, visualid = 827, screen
 = 0, depth = 24, c_class = 4, red_mask = 16711680, green_mask = 65280, blue_mask = 255, colormap_size = 256, bits_per_rgb = 8}
12 glXCreateContext(dpy = 0xaaaae035a560, vis = &{visual = 0xaaaae03687f8, visualid = 827, screen = 0, depth = 24, c_class = 4, red_mask = 16711680, green_mask = 65280, blue_mask = 255, colormap_size = 256, bits_per_rgb = 8}, shareList = NULL, direct = True) = 0xaaaae03b2080
13 glXMakeCurrent(dpy = 0xaaaae035a560, drawable = 58720258, ctx = 0xaaaae03b2080) = True
14 glViewport(x = 0, y = 0, width = 32, height = 32) // fake
15 glScissor(x = 0, y = 0, width = 32, height = 32) // fake
28 glGetIntegerv(pname = GL_NUM_EXTENSIONS, params = &253)
286 glGetIntegerv(pname = GL_NUM_EXTENSIONS, params = &253)
544 glGetIntegerv(pname = GL_NUM_EXTENSIONS, params = &253)
558 glXMakeCurrent(dpy = 0xaaaae035a560, drawable = 0, ctx = NULL) = True
559 glXDestroyContext(dpy = 0xaaaae035a560, ctx = 0xaaaae03b2080)
560 glXChooseVisual(dpy = 0xaaaae035a560, screen = 0, attribList = {GLX_RGBA, GLX_RED_SIZE, 3, GLX_GREEN_SIZE, 3, GLX_BLUE_SIZE, 2, GLX_DOUBLEBUFFER, GLX_DEPTH_SIZE, 16, GLX_X_VISUAL_TYPE, GLX_DIRECT_COLOR, 0}) = &{visual = 0xaaaae036a740, visualid = 970, screen = 0, depth = 24, c_class = 5, red_mask = 16711680, green_mask = 65280, blue_mask = 255, colormap_size = 256, bits_per_rgb = 8}
561 glXCreateContext(dpy = 0xaaaae035a560, vis = &{visual = 0xaaaae036a740, visualid = 970, screen = 0, depth = 24, c_class = 5, red_mask = 16711680, green_mask = 65280, blue_mask = 255, colormap_size = 256, bits_per_rgb = 8}, shareList = NULL, direct = True) = 0xaaaae04911e0
562 glXMakeCurrent(dpy = 0xaaaae035a560, drawable = 58720264, ctx = 0xaaaae04911e0) = True
563 glViewport(x = 0, y = 0, width = 32, height = 32) // fake
564 glScissor(x = 0, y = 0, width = 32, height = 32) // fake
567 glXMakeCurrent(dpy = 0xaaaae035a560, drawable = 0, ctx = NULL) = True
568 glXDestroyContext(dpy = 0xaaaae035a560, ctx = 0xaaaae04911e0)
576 glXChooseVisual(dpy = 0xaaaae035a560, screen = 0, attribList = {GLX_RGBA, GLX_RED_SIZE, 8, GLX_GREEN_SIZE, 8, GLX_BLUE_SIZE, 8, GLX_DOUBLEBUFFER, GLX_DEPTH_SIZE, 24, GLX_STENCIL_SIZE, 8, 0}) = &{visual = 0xaaaae0368948, visualid = 833, screen = 0, depth = 24, c_class = 4, red_mask = 16711680, green_mask = 65280, blue_mask = 255, colormap_size = 256, bits_per_rgb = 8}
581 glXCreateContext(dpy = 0xaaaae035a560, vis = &{visual = 0xaaaae0368948, visualid = 833, screen = 0, depth = 24, c_class = 4, red_mask = 16711680, green_mask = 65280, blue_mask = 255, colormap_size = 256, bits_per_rgb = 8}, shareList = NULL, direct = True) = 0xaaaae04f2f00
582 glXMakeCurrent(dpy = 0xaaaae035a560, drawable = 58720270, ctx = 0xaaaae04f2f00) = True
583 glViewport(x = 0, y = 0, width = 32, height = 32) // fake
584 glScissor(x = 0, y = 0, width = 32, height = 32) // fake
597 glGetIntegerv(pname = GL_NUM_EXTENSIONS, params = &253)
855 glGetIntegerv(pname = GL_NUM_EXTENSIONS, params = &253)
1113 glGetIntegerv(pname = GL_NUM_EXTENSIONS, params = &253)
1127 glXMakeCurrent(dpy = 0xaaaae035a560, drawable = 0, ctx = NULL) = True
1128 glXDestroyContext(dpy = 0xaaaae035a560, ctx = 0xaaaae04f2f00)
1134 glGetIntegerv(pname = GL_MAJOR_VERSION, params = &3)
1135 glGetIntegerv(pname = GL_MINOR_VERSION, params = &1)
1137 glGetIntegerv(pname = GL_NUM_EXTENSIONS, params = &112)
1138 glGetIntegerv(pname = GL_MAX_VERTEX_TEXTURE_IMAGE_UNITS, params = &32)
```

The trace has the answer for your query about `GL_MAJOR_VERSION` and `GL_MINOR_VERSION`, they are 3 and 1 respectively. Although the trace doesn't show `GL_VERSION` being checked, a typical result for that is:

```
OpenGL version string: OpenGL ES 3.1 Mesa 21.3.0-devel (git-580fd7587a)
```

**Comment by icecream95** [comment]
I note that there are no `glGetString` calls in the trace file, so this function might not be reaching the GL driver then?

**Comment by icecream95** [comment]
Ah, no, they are there, I just need to pass `-v` to `apitrace dump`.

For the old GLX contexts, it returns:

```
1110 glGetString(name = GL_VERSION) = "3.3 (Compatibility Profile) Mesa 22.1.0-devel (git-294a357 2022-02-18 impish-oibaf-ppa)"
```

But for the GLES context, the function returns:

```
1129 glGetString(name = GL_RENDERER) = "Mali-G52 (Panfrost)"
1130 glGetString(name = GL_VERSION) = "OpenGL ES 3.1 Mesa 22.1.0-devel (git-294a357 2022-02-18 impish-oibaf-ppa)"
1131 glGetString(name = GL_SHADING_LANGUAGE_VERSION) = "OpenGL ES GLSL ES 3.10"
1132 glGetString(name = GL_VENDOR) = "Panfrost"
1133 glGetError() = GL_NO_ERROR
1134 glGetIntegerv(pname = GL_MAJOR_VERSION, params = &3)
1135 glGetIntegerv(pname = GL_MINOR_VERSION, params = &1)
1136 glGetError() = GL_NO_ERROR
```

```
1152 glGetFloatv(pname = GL_MAX_TEXTURE_MAX_ANISOTROPY_EXT, params = &16)
1153 glGetString(name = GL_VENDOR) = "Panfrost"
1154 glGetString(name = GL_RENDERER) = "Mali-G52 (Panfrost)"
1155 glGetString(name = GL_VERSION) = "OpenGL ES 3.1 Mesa 22.1.0-devel (git-294a357 2022-02-18 impish-oibaf-ppa)"
1156 glGetString(name = GL_SHADING_LANGUAGE_VERSION) = "OpenGL ES GLSL ES 3.10"
1157 glActiveTexture(texture = GL_TEXTURE0)
```

I note that we don't query the extensions again after creating the GLES context, maybe that is related.

**Comment by icecream95** [comment]
The entirety of the trace for context creation (again, EGL functions are not present):

[apitrace-init.txt](https://github.com/hrydgard/ppsspp/files/8107452/apitrace-init.txt)

**Comment by unknownbrackets** [comment]
Oh okay, I think we both made assumptions then.

> > And it seems like we're not detecting your driver as GLES3
> 
> Hey, I didn't write _all_ of it, only a few hundred patches here and there.

Haha, I actually just meant "the driver you're using" - didn't realize you had written a lot of it.

That trace never has EXT_blend_func_extended, but it has ARB_blend_func_extended.  I think we're ending up trying the GL 3.3 extension, and trying dual source blending the GL 3.3 way (which won't work.)  I could force ARB off when we're in GLES mode, but if I'm right, the larger problem is we don't know about any of the correct extensions.

As long as we detect GLES 3 or GL 3, we get the extensions here:
https://github.com/hrydgard/ppsspp/blob/16bf5197e4accf36f4e198a2b099985c0bbb3929/Common/GPU/OpenGL/GLFeatures.cpp#L317-L325

And this is where we check the precision formats (same function):
https://github.com/hrydgard/ppsspp/blob/16bf5197e4accf36f4e198a2b099985c0bbb3929/Common/GPU/OpenGL/GLFeatures.cpp#L487

Which tells me with confidence that this section is that function:

```
1129 glGetString(name = GL_RENDERER) = "Mali-G52 (Panfrost)"
1130 glGetString(name = GL_VERSION) = "OpenGL ES 3.1 Mesa 22.1.0-devel (git-294a357 2022-02-18 impish-oibaf-ppa)"
1131 glGetString(name = GL_SHADING_LANGUAGE_VERSION) = "OpenGL ES GLSL ES 3.10"
1132 glGetString(name = GL_VENDOR) = "Panfrost"
1133 glGetError() = GL_NO_ERROR
1134 glGetIntegerv(pname = GL_MAJOR_VERSION, params = &3)
1135 glGetIntegerv(pname = GL_MINOR_VERSION, params = &1)
1136 glGetError() = GL_NO_ERROR
1137 glGetIntegerv(pname = GL_NUM_EXTENSIONS, params = &112)
1138 glGetIntegerv(pname = GL_MAX_VERTEX_TEXTURE_IMAGE_UNITS, params = &32)
1139 glGetShaderPrecisionFormat(shadertype = GL_VERTEX_SHADER, precisiontype = GL_LOW_FLOAT, range = {15, 15}, precision = {10, 0})
```

But not seeing any `glGetStringi()` for each extension - maybe it's somehow being cached by some layer in between?  We're definitely calling it - and what's more, this function (and therefore that API trace section) is the *only* place we're calling it.

We normally only call this entire function once:
https://github.com/hrydgard/ppsspp/blob/16bf5197e4accf36f4e198a2b099985c0bbb3929/Common/GPU/OpenGL/GLFeatures.cpp#L134-L137

There is a way to reset it...
https://github.com/hrydgard/ppsspp/blob/16bf5197e4accf36f4e198a2b099985c0bbb3929/Common/GPU/OpenGL/GLFeatures.cpp#L578-L579

But it's only used on Windows:
https://github.com/hrydgard/ppsspp/blob/0c40e918c92b897f745abee0d09cf033a1572337/Windows/GPU/WindowsGLContext.cpp#L298-L299

To me this implies that the other GL_EXTENSIONS calls are coming from SDL or similar (unless this is libretro, Qt, or some other port.)  On SDL, we basically try to find the newest context possible:
https://github.com/hrydgard/ppsspp/blob/cf6e11e1d11440b9e97e750b5360d429cc382fcd/SDL/SDLGLGraphicsContext.cpp#L322-L344

For GLES, we'd start with GLES 3.2, and go downward from there.  So I'm not sure how we'd end up with a GL 3.3 context.  We should also reject a surface with 3 bits of red/etc... so I'm wondering if this is Qt or libretro.  In any case, it's definitely a bug somewhere if the GL 3 extensions are getting cached and returned by `glGetStringi()`...

-[Unknown]

**Comment by icecream95** [comment]
I'm looking at the SDL2 source code.

There is no separate SDL video driver for GLES on X11, so it starts by going through the normal GLX context create path, which creates a desktop OpenGL context.

```
            context = _this->gl_data->glXCreateContext(display, vinfo,
                                                        NULL, True);
```

SDL then checks for supported OpenGL ES compatibility extensions:

```
void
SDL_GL_DeduceMaxSupportedESProfile(int* major, int* minor)
{
/* THIS REQUIRES AN EXISTING GL CONTEXT THAT HAS BEEN MADE CURRENT. */
/*  Please refer to https://bugzilla.libsdl.org/show_bug.cgi?id=3725 for discussion. */
#if SDL_VIDEO_OPENGL || SDL_VIDEO_OPENGL_ES || SDL_VIDEO_OPENGL_ES2
    /* XXX This is fragile; it will break in the event of release of
     * new versions of OpenGL ES.
     */
    if (SDL_GL_ExtensionSupported("GL_ARB_ES3_2_compatibility")) {
        *major = 3;
        *minor = 2;
    } else if (SDL_GL_ExtensionSupported("GL_ARB_ES3_1_compatibility")) {
        *major = 3;
        *minor = 1;
    } else if (SDL_GL_ExtensionSupported("GL_ARB_ES3_compatibility")) {
        *major = 3;
        *minor = 0;
    } else {
        *major = 2;
        *minor = 0;
    }
#endif
}
```

Panfrost does not support `GL_ARB_ES3_1_compatibility` due to not having `ARB_gpu_shader5` nor `ARB_compute_shader` yet, though we do support compute shaders in GLES.

```
    /* If we need a GL ES context and there's no  
     * GLX_EXT_create_context_es2_profile extension, switch over to X11_GLES functions  
     */
    if (((_this->gl_config.profile_mask == SDL_GL_CONTEXT_PROFILE_ES) ||
         SDL_GetHintBoolean(SDL_HINT_VIDEO_X11_FORCE_EGL, SDL_FALSE)) &&
        X11_GL_UseEGL(_this) ) {
```

`X11_GL_UseEGL` will return `true` because we do not support the required compatibility extension, and so `X11_GLES_LoadLibrary` is called, which uses `SDL_EGL_CreateContext` to create a GLES context with EGL.

<hr>

I've realised that `gl_extensions.GLES3` *is* getting set after all..

```
        } else if (gl_extensions.IsGLES && gl_extensions.GLES3) {
                version = StringFromFormat("#version %d es\n", gl_extensions.GLSLVersion());
        }
```

Shaders have `#version 310 es` lines, and so `GLES3` must be getting set.

<hr>

So I think that `glBindFragDataLocationIndexedEXT` *is* getting called after all, just there is some bug in Apitrace that makes it not trace the function.

But… the glitched rending is something that would *only* happen when `glBindFragDataLocationIndexedEXT` is not called, or if it doesn't work for some other reason. The dual-source rendering bug that I fixed yesterday only caused GPU faults (which show up as half-rendered frames); it wouldn't cause artefacts like the one in the reports.

…Or maybe it would; I have a different GPU model (Mali G72) to the other people who have reported the issue (Mali G52). If the Mali G52 doesn't raise faults in this situation, then I guess that it could confuse the internal state of the GPU and result in something similar to what I saw.

The user who gave me the Apitrace reports that with the bug fix for Panfrost, PPSSPP does render correctly with no other changes. So probably there is no problem with PPSSPP then? Sorry for wasting your time, I really need to fix my habit of seeing something suspicious and then assuming that it must be the cause of some issue…

**Comment by unknownbrackets** [comment]
Ah okay, so maybe it wasn't being cached anyway.  Thanks for the rundown - glad it was something simpler, after all.  I did add some asserts that may prevent confusion in the future, so it wasn't for nothing.

We run into a lot of driver bugs across the set, and some of them are still open and it can be hard to engage closed source driver authors.  Always nice to see such a bug one fixed - they're tricky beasts.

-[Unknown]
